components/datetime/src/builder.rs

```diff
@@ -0,0 +1,236 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::fieldsets::{self, enums::*};
+use crate::options::*;
+
+/// An enumeration over all possible date and calendar period field sets
+/// without options.
+///
+/// This is a builder enum. See [`builder`](crate::fieldsets::builder).
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DateFields {
+    /// The day of the month, as in
+    /// “on the 1st”.
+    D,
+    /// The month and day of the month, as in
+    /// “January 1st”.
+    MD,
+    /// The year, month, and day of the month, as in
+    /// “January 1st, 2000”.
+    YMD,
+    /// The day of the month and day of the week, as in
+    /// “Saturday 1st”.
+    DE,
+    /// The month, day of the month, and day of the week, as in
+    /// “Saturday, January 1st”.
+    MDE,
+    /// The year, month, day of the month, and day of the week, as in
+    /// “Saturday, January 1st, 2000”.
+    YMDE,
+    /// The day of the week alone, as in
+    /// “Saturday”.
+    E,
+    /// A standalone month, as in
+    /// “January”.
+    M,
+    /// A month and year, as in
+    /// “January 2000”.
+    YM,
+    /// A year, as in
+    /// “2000”.
+    Y,
+}
+
+/// An enumeration over all possible time zone styles.
+///
+/// This is a builder enum. See [`builder`](crate::fieldsets::builder).
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ZoneStyle {
+    /// The long specific non-location format, as in
+    /// “Pacific Daylight Time”.
+    Z,
+    /// The short specific non-location format, as in
+    /// “PDT”.
+    Zs,
+    /// The long offset format, as in
+    /// “GMT−8:00”.
+    O,
+    /// The short offset format, as in
+    /// “GMT−8”.
+    Os,
+    /// The long generic non-location format, as in
+    /// “Pacific Time”.
+    V,
+    /// The short generic non-location format, as in
+    /// “PT”.
+    Vs,
+    /// The location format, as in
+    /// “Los Angeles time”.
+    L,
+}
+
+#[derive(Debug, displaydoc::Display)]
+#[non_exhaustive]
+pub enum BuilderError {
+    /// The specified fields are incompatible with the desired field set
+    InvalidFields,
+    /// The specified options are incompatible with the specified field set
+    InvalidOptions,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
+#[non_exhaustive]
+pub struct FieldSetBuilder {
```

Comment:
docs nit: should have docs explaining use, and talking qualitatively about which combinations are valid. users should ideally not have to use trial and error to figure out what fields they need to set for their use case